File: examples/python/oha/cmd/delete.py
#!/usr/bin/python3


"""Command ``delete`` is used to delete one or more records from the db.

It can repeat the same operation in loop or a defined numbers of times. It can
be possible to define a delay between one operation and the next one and also
after how many operations commit the changes.
"""


import time
import click

import cx_Oracle


@click.command()
@click.option('--loop/--no-loop',
              type=bool,
              default=False,
              help="repeat the same operation forever")
@click.option('--iters',
              type=int,
              default=1,
              help='repeat the same operation a given number of times')
@click.option('--delay',
              type=float,
              default=0.25,
              help='time to wait before the next iteration')
@click.option('--commit-every',
              type=int,
              default=1,
              help='after how many operations perform a commit')
@click.pass_obj
def delete(obj,
           loop: bool,
           iters: int,
           delay: float,
           commit_every: int) -> None:
    """Delete records from the table"""

    # Define query parameters
    table = obj.conf["database"]["table"]
    conds = {"id": 0}

    iters = 0 if loop else iters
    step = 1
    try:
        while loop or step <= iters:
            # Prepare query with updated conditions
            pairs = [f"{cond}={val + step}" for cond, val in conds.items()]
            query = f"DELETE FROM {table} WHERE {', '.join(pairs)}"

            # Execute query
            try:
                obj.cur.execute(query)
                click.echo(f"[{step}/{iters}] - {query}")
            except cx_Oracle.IntegrityError as err:
                click.echo(err)
                break

            # Commit changes
            if step % commit_every == 0:
                obj.conn.commit()
                click.echo(f"[{step}/{iters}] - COMMIT")

            step += 1
            time.sleep(delay)
    except KeyboardInterrupt as _:
        click.echo("Error - Interrupted by the user")

    # Check the last commit
    if iters % commit_every != 0:
        obj.conn.commit()
        click.echo(f"[{iters}/{iters}] - COMMIT")

File: examples/golang/oracle-ha/cmd/delete.go
/*
Copyright © 2022 Andrea Lauri <andrea.lauri86@gmail.com>

Command ``delete`` is used to delete one or more records from the db.

It can repeat the same operation in loop or a defined numbers of times. It can
be possible to define a delay between one operation and the next one and also
after how many operations commit the changes.
*/
package cmd

import (
    "time"
	"fmt"

	"github.com/spf13/cobra"
	"github.com/spf13/viper"
)

type conds struct {
    id int
}

// deleteCmd represents the delete command
var deleteCmd = &cobra.Command{
	Use:   "delete",
	Short: "Delete records from the table.",
	Long: `Delete records from the table.`,
	Run: func(cmd *cobra.Command, args []string) {
        loop, _ := cmd.Flags().GetBool("loop")
        iters, _ := cmd.Flags().GetInt("iters")
        delay, _ := cmd.Flags().GetFloat64("delay")
        commit_every, _ := cmd.Flags().GetInt("commit-every")
        table := viper.GetViper().GetString("database.table")
        conditions := conds{0}

        if loop { iters = 0 }
        step := 1
        for {
            if !loop && step > iters {
                break
            }

            pairs := conds{conditions.id + step}
            query := fmt.Sprintf("DELETE FROM %s WHERE id=%d", table, pairs.id)

            // TODO: replace with call to cx_Oracle driver
            fmt.Fprintln(cmd.OutOrStdout(),
                         fmt.Sprintf("[%d/%d] - %s", step, iters, query))

            // Commit changes
            if step % commit_every == 0 {
                // TODO: insert a database commit operation here
                fmt.Fprintln(cmd.OutOrStdout(),
                             fmt.Sprintf("[%d/%d] - COMMIT", step, iters))
            }

            step += 1
            time.Sleep(time.Duration(delay) * time.Millisecond)
        }

        // Chech the last commit
        if iters % commit_every != 0 {
            // TODO: insert a database commit operation here
            fmt.Fprintln(cmd.OutOrStdout(),
                         fmt.Sprintf("[%d/%d] - COMMIT", iters, iters))
        }
	},
}

func init() {
    deleteCmd.Flags().Bool("loop", false, "repeat the same operation forever")
    deleteCmd.Flags().Int("iters", 1, "repeat the same operation a given number of times")
    deleteCmd.Flags().Float64("delay", 0.25, "time to wait before the next iteration")
    deleteCmd.Flags().Int("commit-every", 1, "after how many operations perform a commit")
}
